### PR TITLE
Add contribution of MB to gradient computation

### PR DESCRIPTION
**Pull request by albangossard**
Companion PRs:
- https://github.com/ODINN-SciML/Sleipnir.jl/pull/134
- https://github.com/ODINN-SciML/Muninn.jl/pull/32

Fixes https://github.com/ODINN-SciML/ODINN.jl/issues/365

### CODE REVIEW

**Review by JordiBolibar — COMMENTED** [review]
No comments on my side, but I let @facusapienza21 revise the internals of the gradient computation.

File: src/inverse/SIA2D/gradient.jl [hunk 1]
```diff
@@ -120,13 +132,13 @@ function SIA2D_grad_batch!(θ, simulation::FunctionalInversion)
                 ℓ += Δt[j-1]*ℓi
 
                 ### Custom VJP to compute the adjoint
-                λ_∂f∂H, dH_H = VJP_λ_∂SIA∂H(simulation.parameters.UDE.grad.VJP_method, λ[j], H[j], θ, simulation, t₀)
+                λ_∂f∂H, dH_H = VJP_λ_∂SIA∂H(simulation.parameters.UDE.grad.VJP_method, λ[j], H[j], θ, simulation, tj)
```
Comment (facusapienza21) [inline src/inverse/SIA2D/gradient.jl:135]: Was this a bug in the original code? The fact that this was calling the time t0 rather that then current time? I think this probably had no effect before, since none of the dynamics depended of time before (with MB, this is no longer true)

File: src/inverse/SIA2D/gradient.jl [hunk 2]
```diff
@@ -157,44 +169,63 @@ function SIA2D_grad_batch!(θ, simulation::FunctionalInversion)
...
+                    t = - integrator.t
+                    ∂ℓ∂H, ∂ℓ∂θ = backward_loss(
+                        loss_function,
+                        H_itp(t),
+                        H_ref_itp(t),
+                        t,
+                        glacier,
+                        θ,
+                        simulation;
+                        normalization = prod(N) * normalization
```
Comment (facusapienza21) [inline src/inverse/SIA2D/gradient.jl:200]: @albangossard what is the logic behing doing this inside a `let` block?

File: src/inverse/SIA2D/gradient.jl [hunk 3]
```diff
@@ -157,44 +169,63 @@ function SIA2D_grad_batch!(θ, simulation::FunctionalInversion)
...
+            cb_adjoint_MB = if simulation.parameters.simulation.use_MB
+                nSteps = Int(round((tspan[2]-tspan[1])/simulation.parameters.solver.step))
+                tstopsMB = - (tspan[1] .+ collect(1:nSteps) .* simulation.parameters.solver.step)
+                stop_condition_MB(λ, t, integrator) = Sleipnir.stop_condition_tstops(λ, t, integrator, tstopsMB)
+                DiscreteCallback(stop_condition_MB, effect_MB!)
+            else
+                CallbackSet()
             end
-            cb_adjoint_loss = DiscreteCallback(stop_condition, effect!)
+            cb = CallbackSet(cb_adjoint_MB, cb_adjoint_loss)
```
Comment (facusapienza21) [inline src/inverse/SIA2D/gradient.jl:221]: Maybe worth mentioning somewhere that the tstops of these two callbacks don't need to be the same! I think it still works when MB and losses are aggregated at different time, but just worth checking this!

File: src/inverse/SIA2D/VJPs.jl [hunk 4]
```diff
@@ -67,3 +68,41 @@ function VJP_λ_∂surface_V∂θ(VJPMode::DiscreteVJP, λx, λy, H, θ, simulat
...
+    MB = Enzyme.make_zero(H)
+    λH = deepcopy(λ) # Need to copy because Enzyme changes the backward gradient in-place
+    Enzyme.autodiff(
+        Reverse, MB_wrapper!, Const,
+        Duplicated(MB, λH),
+        Duplicated(H, λ_∂MB∂H),
+        Duplicated(simulation, _simulation),
+        Duplicated(glacier, _glacier),
+    )
+    return λ + λ_∂MB∂H
```
Comment (facusapienza21) [inline src/inverse/SIA2D/VJPs.jl:103]: I find a bit confusing that this returns the adjoint itself and the contribution of MB at the same time. I left another comment in `gradient.jl` regarding this point, but wouldn't make more sente to return just `λ_∂MB∂H` and then in the gradient calculation have a `.+=` instead of a `.=`?

File: src/inverse/SIA2D/gradient.jl [hunk 5]
```diff
@@ -157,44 +169,63 @@ function SIA2D_grad_batch!(θ, simulation::FunctionalInversion)
...
-                integrator.u .= integrator.u .+ simulation.parameters.simulation.step .* ∂ℓ∂H
+                    integrator.u .= integrator.u .+ simulation.parameters.simulation.step .* ∂ℓ∂H
+                end
+            end
+            cb_adjoint_loss = DiscreteCallback(stop_condition_loss, effect_loss!)
+            effect_MB! = let simulation=simulation, glacier=glacier, H_itp=H_itp
+                function (integrator)
+                    t = - integrator.t
+                    λ_∂MB∂H = VJP_λ_∂MB∂H(simulation.parameters.UDE.grad.MB_VJP, integrator.u, H_itp(t), simulation, glacier, t)
+                    integrator.u .= λ_∂MB∂H
```
Comment (facusapienza21) [inline src/inverse/SIA2D/gradient.jl:210]: Also following my comment before. I think it makes more sense to do `integrator.u .+= ` and just have the contribution of the adjoint. Am I missing something Alban?

File: src/inverse/VJPTypes.jl [hunk 6]
```diff
@@ -34,4 +34,12 @@ Enzyme implementation of VJP used inside the adjoint calculation.
...
\ No newline at end of file
+end
+
+"""
+No VJP flavor when the contribution of a given term should not be computed inside the adjoint calculation (e.g. MB).
+
+`NoVJP`
+"""
+@kwdef struct NoVJP <: AbstractVJPMethod
+end
```
Comment (facusapienza21) [inline src/inverse/VJPTypes.jl:45]: I like it!

File: src/inverse/SIA2D/VJPs.jl [hunk 7]
```diff
@@ -67,3 +68,41 @@ function VJP_λ_∂surface_V∂θ(VJPMode::DiscreteVJP, λx, λy, H, θ, simulat
...
+    λ_∂MB∂H = Enzyme.make_zero(H)
+    MB = Enzyme.make_zero(H)
+    λH = deepcopy(λ) # Need to copy because Enzyme changes the backward gradient in-place
+    Enzyme.autodiff(
+        Reverse, MB_wrapper!, Const,
+        Duplicated(MB, λH),
+        Duplicated(H, λ_∂MB∂H),
+        Duplicated(simulation, _simulation),
+        Duplicated(glacier, _glacier),
+    )
```
Comment (facusapienza21) [inline src/inverse/SIA2D/VJPs.jl:102]: I think is great this is super generic and it should work for any MB model, but notice that the adjoint contribution of the mass balance is actually very simple to compute and it is given by $\frac{\partial MB}{\partial H} \lambda$. For very simple MBs (like the ones we are using), the partial derivative $frac{\partial MB}{\partial H}$ is trivial to compute. This can be the default for some simple MBs models, but I imagine we can have at least one MB where we do this and we just use it for testing. Does this makes sense? Maybe it is a bit too much, but the actual contribution of MB to the adjoint is mathematically very simple, reason why I think we should consider having this done more "manually". What do you think @albangossard ?

**Review by facusapienza21 — COMMENTED** [review]
@albangossard I left a few comments and questions. Overall, all these looks great! I just want to iterate on a few items I left in this PR.
